src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,276 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using System.Collections.Generic;
-using System.Diagnostics;
+using System.Buffers;
+using System.Buffers.Binary;
 using System.IO;
 using System.IO.Pipelines;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.Serialization;
-using Nerdbank.Streams;
 using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.Remote
+namespace Microsoft.CodeAnalysis.Remote;
+
+/// <summary>
+/// Contains the utilities for writing assets from the host to a pipe-writer and for reading those assets on the
+/// server.  The format we use is as follows.  For each asset we're writing we write:
+/// <code>
+/// -------------------------------------------------------------------------
+/// | sentinel (1 byte) | length of data (4 bytes) | data (variable length) |
+/// -------------------------------------------------------------------------
+/// </code>
+/// The writing code will write out the sentinel-byte and data-length, ensuring it is flushed to the pipe-writer.
+/// This allows the pipe-reader to immediately read that information so it can then pre-allocate the space for the
+/// data to go into. After writing the data the writer will also flush, so the reader can then read the data out of
+/// the pipe into its buffer.  Once present in the reader's buffer, synchronous deserialization can happen without
+/// any sync-over-async blocking on async-io.
+/// <para>
+/// The sentinel byte serves to let us detect immediately on the reading side if something has gone wrong with this
+/// system.
+/// </para>
+/// <para>
+/// In order to be able to write out the data-length, the writer will first synchronously write the asset to an
+/// in-memory buffer, then write that buffer's length to the pipe-writer, then copy the in-memory buffer to the
+/// writer.
+/// </para>
+/// When writing/reading the data-segment, we use an the <see cref="ObjectWriter"/>/<see cref="ObjectReader"/>
+/// subsystem.  This will write its own validation bits, and then the data describing the asset.  This data is:
+/// <code>
+/// -----------------------------------------------------------------------------------------------------------
+/// | data (variable length)                                                                                  |
+/// -----------------------------------------------------------------------------------------------------------
+/// | ObjectWriter validation (2 bytes) | checksum (16 bytes) | kind (4 bytes) | asset-data (asset specified) |
```

Comment:
>kind (4 bytes) [](http://example.com/codeflow?start=64&length=14)

nit: I feel like making this enum byte sized would still give it enough room to grow since there are currently only 15 items in it

Reply:
Sure!